google-cloud-spanner/src/main/java/com/google/cloud/spanner/MultiplexedSessionDatabaseClient.java

```diff
@@ -0,0 +1,428 @@
+              || spannerException instanceof InstanceNotFoundException
+              || spannerException instanceof SessionNotFoundException)) {
+        // This could in theory set this field more than once, but we don't want to bother with
+        // synchronizing, as it does not really matter exactly which error is set.
+        this.client.resourceNotFoundException.set((ResourceNotFoundException) spannerException);
+      }
+    }
+
+    @Override
+    void onReadDone() {
+      // This method is called whenever a ResultSet that was returned by this transaction is closed.
+      // Close the active transaction if this is a single-use transaction. This ensures that the
+      // active span is ended.
+      if (this.singleUse && getActiveTransaction() != null) {
+        getActiveTransaction().close();
+        setActive(null);
+        if (this.singleUseChannelHint != NO_CHANNEL_HINT) {
+          this.client.channelUsage.clear(this.singleUseChannelHint);
+        }
+        this.client.numCurrentSingleUseTransactions.decrementAndGet();
+      }
+    }
+
+    @Override
+    public void close() {
+      // no-op, we don't want to delete the multiplexed session.
+    }
+  }
+
+  /**
+   * Keeps track of which channels have been 'given' to single-use transactions for a given Spanner
+   * instance.
+   */
+  private static final Map<SpannerImpl, BitSet> CHANNEL_USAGE = new HashMap<>();
+
+  private final BitSet channelUsage;
+
+  private final int numChannels;
+
+  /**
+   * The number of single-use read-only transactions currently running on this multiplexed session.
+   */
+  private final AtomicInteger numCurrentSingleUseTransactions = new AtomicInteger();
+
+  private boolean isClosed;
+
+  /** The duration before we try to replace the multiplexed session. The default is 7 days. */
+  private final Duration sessionExpirationDuration;
+
+  private final SessionClient sessionClient;
+
+  private final TraceWrapper tracer;
+
+  /** The current multiplexed session that is used by this client. */
+  private final AtomicReference<ApiFuture<SessionReference>> multiplexedSessionReference;
+
+  /** The expiration date/time of the current multiplexed session. */
+  private final AtomicReference<Instant> expirationDate;
+
+  /**
+   * The maintainer runs every 10 minutes to check whether the multiplexed session should be
+   * refreshed.
+   */
+  private final MultiplexedSessionMaintainer maintainer;
+
+  /**
+   * If a {@link DatabaseNotFoundException} or {@link InstanceNotFoundException} is returned by the
+   * server, then we set this field to mark the client as invalid.
+   */
+  private final AtomicReference<ResourceNotFoundException> resourceNotFoundException =
+      new AtomicReference<>();
+
+  /**
+   * This flag is set to true if the server return UNIMPLEMENTED when we try to create a multiplexed
+   * session.
+   */
+  private final AtomicBoolean unimplemented = new AtomicBoolean(false);
```

Comment:
Nit: The formatting looks off. Also, should we add a TODO that this is a temporary measure and we can remove it in future when this has a bit of increased usage?

Reply:
I added a TODO for removing.

I'm not sure what you mean with the formatting being off (?) The code is formatted using the normal formatter, and I don't see anything strange with this line or the ones directly above.